### ISSUE TITLE
Cluster down kine.sock connection refused

### ISSUE THREAD

**Issue by JasonPulse** [canonical/microk8s#4523]
#### Summary
Came home to have my 2 node cluster down with both nodes reporting the error below for kubelite service
`May 16 06:11:30 ubuntu microk8s.daemon-kubelite[6065]: W0516 06:11:30.766110    6065 logging.go:59] [core] [Channel #7 SubChannel #9] grpc: addrConn.createTransport failed to connect to {Addr: "unix:///var/snap/microk8s/6787/var/kubernetes/backend/kine.sock:12379", ServerName: "kine.sock:12379", }. Err: connection error: desc = "transport: Error while dialing: dial unix /var/snap/microk8s/6787/var/kubernetes/backend/kine.sock:12379: connect: connection refused"`

#### What Should Happen Instead?
Running cluster

#### Reproduction Steps
Unknown at this time, assumed something to do with a snap refresh.

#### Introspection Report
System is unstable 20-30 seconds to echo back typing with microk8s trying to start, tar is with microk8s stopped, if requested I can provide files from /var/snap/microk8s/current/var/kubernets/backend as there is no sensitive data running on cluster

Would like to know why this happened as we have several clusters running the same things and with them all failing for a different issue of nf_conntrack having another issue to go back to will not be fun.

[inspection-report-20240516_151937.tar.gz](https://github.com/canonical/microk8s/files/15337426/inspection-report-20240516_151937.tar.gz)

**Comment by JasonPulse** [comment]
After about 7 hours I was able to get a Tar from running microk8s
[inspection-report-20240516_223539.tar.gz](https://github.com/canonical/microk8s/files/15343648/inspection-report-20240516_223539.tar.gz)

**Comment by mikhatanu** [comment]
just encountered this error too, need help.

**Comment by JasonPulse** [comment]
Yeah no one ever answered or helped here, I've had this issue many times since.  In this case it was just faster to blow away the cluster and rebuild then to recover.  I have noticed this is usually related to the host system writing to the HDD I think in this case the main node was a Pi with a memory card as main HDD and that was just not fast enough.  Another time I have received this is the node running out of HDD space or damn near close to it.

You can edit /var/snap/microk8s/current/args/kubelet and add the -gc threshold flags to change the image pruning for a more reasonable tolerance then 85%

```
--image-gc-high-threshold=60
--image-gc-low-threshold=30
```

**Comment by AlexGustafsson** [comment]
I'm seeing these error logs on 1.35 as well, once every ~10s, but I haven't noticed any issues. It's only on one node as well.

**Comment by susannaweinberger** [comment]
Hi - we also see the problem on 1.34 and 1.36 (but everything is working fine as we see)
Anyone knows what the problem is?

**Comment by yenneferofvengerberg** [comment]
We still have the same issue, and it appears randomly. In our scenario, the failure triggers a cascading failure on other services, including on calico. Additionally the leader does not properly reconcile, which causes calico pods not being rescheduled. The attached script, tries to heal the cluster, and in our case, it seems to work. Sometimes it restarts dqlite more than necessary, but I couldn't infer the precise condition when it should be restarted. The script has been vetted for publication, however, please, validate it before running since uses our defaults (for example, the build namespace). Our clusters can be easily rebuilt from scratch in case of issues, but this time I didn't want to go ballistic :)

[microk8s-healer-v10.5.sh](https://github.com/user-attachments/files/30872479/microk8s-healer-v10.5.sh)
